File: framework-java/src/main/java/org/android/framework/util/JetPackUtil.java
package org.android.framework.util;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelStoreOwner;
import androidx.viewbinding.ViewBinding;

import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

public class JetPackUtil {

    private static final ViewModelProvider.Factory FACTORY = new ViewModelProvider.NewInstanceFactory();

    public static <VM extends ViewModel> VM getVM(ViewModelStoreOwner viewModelStoreOwner, Class<VM> vmClazz) {
        return new ViewModelProvider(viewModelStoreOwner, FACTORY).get(vmClazz);
    }

    public static <VB extends ViewBinding> VB getVB(Activity activity) {
        try {
            Class vbClazz = findViewBindingClass(activity);
            Method[] methods = vbClazz.getDeclaredMethods();
            for (Method method : methods) {
                Class<?>[] params = method.getParameterTypes();
                if (params.length == 1 && params[0] == LayoutInflater.class) {
                    return (VB) method.invoke(null, LayoutInflater.from(activity));
                }
            }
        } catch (Exception ignore) {}
        throw new RuntimeException(activity.getClass().getName() + " viewbinding can't find");
    }

    public static <VB extends ViewBinding> VB getVB(Fragment fragment, LayoutInflater layoutInflater, ViewGroup viewGroup) {
        try {
            Class vbClazz = findViewBindingClass(fragment);
            Method[] methods = vbClazz.getDeclaredMethods();
            for (Method method : methods) {
                Class<?>[] params = method.getParameterTypes();
                if (params.length == 3 && params[0] == LayoutInflater.class && params[1] == ViewGroup.class) {
                    return (VB) method.invoke(null, layoutInflater, viewGroup, false);
                }
            }
        } catch (Exception ignore) {}
        throw new RuntimeException(fragment.getClass().getName() + " viewbinding can't find");
    }

    private static Class findViewBindingClass(Object o) throws ClassNotFoundException {
        for (Class clazz = o.getClass(); clazz != null; clazz = clazz.getSuperclass()) {
            Type classType = clazz.getGenericSuperclass();
            if (!(classType instanceof ParameterizedType)) continue;

            Type[] types = ((ParameterizedType) classType).getActualTypeArguments();
            if (types.length < 1) continue;

            Class vbClazz = null;
            for (Type genericType : types) {
                if (ViewBinding.class.isAssignableFrom((Class<?>) genericType)) {
                    vbClazz = (Class) genericType;
                    break;
                }
            }
            if (vbClazz != null) return vbClazz;
        }
        throw new ClassNotFoundException();
    }
}

File: framework/src/main/java/org/android/framework/util/JetPackUtil.kt
package org.android.framework.util

import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelStoreOwner
import androidx.viewbinding.ViewBinding
import java.lang.reflect.ParameterizedType

@Suppress("UNCHECKED_CAST")
object JetPackUtil {

    private val FACTORY: ViewModelProvider.Factory = ViewModelProvider.NewInstanceFactory()

    fun <VM : ViewModel> getVM(
        viewModelStoreOwner: ViewModelStoreOwner,
        vmClazz: Class<VM>
    ): VM {
        return ViewModelProvider(viewModelStoreOwner, FACTORY).get(vmClazz)
    }

    fun <VM : ViewModel> getVM(viewModelStoreOwner: ViewModelStoreOwner): VM {
        try {
            val vmClazz: Class<VM> = findViewModelClass(viewModelStoreOwner) as Class<VM>
            return getVM(viewModelStoreOwner, vmClazz)
        } catch (ignore: Exception) {}
        throw RuntimeException(viewModelStoreOwner.javaClass.getName() + " viewmodel can't find")
    }

    fun <VB : ViewBinding?> getVB(activity: Activity): VB {
        try {
            val vbClazz: Class<*> = findViewBindingClass(activity)
            val methods = vbClazz.declaredMethods
            for (method in methods) {
                val params = method.parameterTypes
                if (params.size == 1 && params[0] == LayoutInflater::class.java) {
                    return method.invoke(null, LayoutInflater.from(activity)) as VB
                }
            }
        } catch (ignore: Exception) {}
        throw RuntimeException(activity.javaClass.getName() + " viewbinding can't find")
    }

    fun <VB : ViewBinding?> getVB(
        fragment: Fragment,
        layoutInflater: LayoutInflater,
        viewGroup: ViewGroup?
    ): VB {
        try {
            val vbClazz: Class<*> = findViewBindingClass(fragment)
            val methods = vbClazz.declaredMethods
            for (method in methods) {
                val params = method.parameterTypes
                if (params.size == 3 && params[0] == LayoutInflater::class.java && params[1] == ViewGroup::class.java) {
                    return method.invoke(null, layoutInflater, viewGroup, false) as VB
                }
            }
        } catch (ignore: Exception) {}
        throw RuntimeException(fragment.javaClass.getName() + " viewbinding can't find")
    }

    private fun findViewModelClass(any: Any): Class<*> {
        var clazz: Class<*>? = any.javaClass
        while (clazz != null) {
            val classType = clazz.genericSuperclass
            if (classType !is ParameterizedType) {
                clazz = clazz.superclass
                continue
            }

            val types = classType.actualTypeArguments
            if (types.isEmpty()) {
                clazz = clazz.superclass
                continue
            }

            var vmClazz: Class<ViewModel>? = null
            for (type in types) {
                if (ViewModel::class.java.isAssignableFrom((type as Class<*>))) {
                    vmClazz = type as Class<ViewModel>
                    break
                }
            }
            if (vmClazz == null) {
                clazz = clazz.superclass
                continue
            }

            return vmClazz
        }
        throw ClassNotFoundException()
    }

    private fun findViewBindingClass(any: Any): Class<*> {
        var clazz: Class<*>? = any.javaClass
        while (clazz != null) {
            val classType = clazz.genericSuperclass
            if (classType !is ParameterizedType) {
                clazz = clazz.superclass
                continue
            }

            val types = classType.actualTypeArguments
            if (types.isEmpty()) {
                clazz = clazz.superclass
                continue
            }

            var vbClazz: Class<ViewBinding>? = null
            for (type in types) {
                if (ViewBinding::class.java.isAssignableFrom((type as Class<*>))) {
                    vbClazz = type as Class<ViewBinding>
                    break
                }
            }
            if (vbClazz == null) {
                clazz = clazz.superclass
                continue
            }

            return vbClazz
        }
        throw ClassNotFoundException()
    }
}